prereise/gather/demanddata/transportation_electrification/smart_charging.py

```diff
@@ -0,0 +1,365 @@
+                    charging_consumption = (
+                        []
+                    )  # zero the consumption, this is a vector, the consumption for each trip
+                    rates = (
+                        []
+                    )  # zero the rates of all segments, which will be used as cost function later
+                    elimit = (
+                        []
+                    )  # zero the charging energy limits of all segments, which will be used as the upper bound
+                    seg = []
+
+                    # total_trips * 2 matrix, recording dwelling start and end time.
+                    dwelling_times = []
+
+                    while trip_num < total_trips:
+                        # DC consumption
+                        charging_consumption.append(
+                            individual.iloc[
+                                trip_num, individual.columns.get_loc("Miles traveled")
+                            ]
+                            * kwhmi
+                            * -1
+                        )
+
+                        # dwelling part, energylimit is the charging energy constraint
+                        # get segment, t1, tn, s1, sn, energylimit, extracost from dwell function
+                        dwelling_start_time = individual.iloc[
+                            trip_num,
+                            individual.columns.get_loc("End time (hour decimal)"),
+                        ]
+                        dwelling_length = individual.iloc[
+                            trip_num,
+                            individual.columns.get_loc("Dwell time (hour decimal)"),
+                        ]
+                        total_dwell_period = dwelling_start_time + dwelling_length
+
+                        segment = dwelling.get_segment(
+                            dwelling_start_time, total_dwell_period
+                        )
+                        power = dwelling.get_charging_power(
+                            power,
+                            trip_strategy,
+                            location_strategy,
+                            individual.iloc[trip_num],
+                        )
+                        energylimit = dwelling.get_energy_limit(
+                            power,
+                            segment,
+                            dwelling_start_time,
+                            dwelling_length,
+                            total_dwell_period,
+                        )
+
+                        # get the rates and the extra cost related to the infrasctructures; used as cost function
+                        for _ in range(dwelling_start_time, dwelling_length + 1):
+                            rates.append(cost[_])
+
+                        # get the energy limit; used as constraint
+                        for s in range(segment):
+                            elimit.append(energylimit)
```

Comment:
Is this equivalent to `elimit += [energylimit] * s`?

Reply:
done